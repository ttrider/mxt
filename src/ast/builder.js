"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const DomUtils_1 = require("DomUtils");
const t = __importStar(require("@babel/types"));
function isBuilder(value) {
    return (value && value.build);
}
function statementList(statements, ...node) {
    const items = (statements === undefined) ? [] : statements;
    const data = {
        statements: items,
        add
    };
    return add(...node);
    function add(...node) {
        for (let statement of node) {
            if (isBuilder(statement)) {
                statement = statement.build();
            }
            if (t.isStatement(statement)) {
                items.push(statement);
            }
            else if (t.isExpression(statement)) {
                items.push(t.expressionStatement(statement));
            }
            else {
                if (statement.statements) {
                    items.push(...statement.statements);
                }
            }
        }
        return data;
    }
}
exports.statementList = statementList;
function declareFunction(functionName) {
    const parameters = [];
    const body = statementList();
    let isExport = false;
    const context = {
        get export() { isExport = true; return context; },
        param: addParam,
        body: addStatement,
        build: buildStatement
    };
    return context;
    function buildStatement() {
        // identifiy optional parameters
        let canBeOptional = true;
        for (let i = parameters.length - 1; i >= 0; i--) {
            if (!canBeOptional) {
                parameters[i].optional = false;
            }
            else {
                if (!parameters[i].optional) {
                    canBeOptional = false;
                }
            }
        }
        if (functionName) {
            const func = t.functionDeclaration(t.identifier(functionName), parameters, t.blockStatement(body.statements));
            if (isExport) {
                return t.exportNamedDeclaration(func, []);
            }
            return func;
        }
        return t.arrowFunctionExpression(parameters, t.blockStatement(body.statements));
    }
    function addStatement(...node) {
        body.add(...node);
        return context;
    }
    function addParam(name, ...types) {
        const i = t.identifier(name);
        if (types.length === 0) {
            i.typeAnnotation = t.tsTypeAnnotation(t.tsAnyKeyword());
        }
        else if (types.length === 1) {
            i.typeAnnotation = t.tsTypeAnnotation(getType(types[0]));
        }
        else {
            i.typeAnnotation = t.tsTypeAnnotation(t.tsUnionType(types.map(getType)));
        }
        parameters.push(i);
        return context;
        function getType(tp) {
            if (tp === undefined) {
                i.optional = true;
                return (t.tsUndefinedKeyword());
            }
            if (tp === null) {
                return (t.tsNullKeyword());
            }
            return (t.tsTypeReference(t.identifier(tp)));
        }
    }
}
exports.declareFunction = declareFunction;
function declareVar(name) {
    let kind = "const";
    let initExpression;
    const context = {
        get const() { kind = "const"; return context; },
        get let() { kind = "let"; return context; },
        get var() { kind = "var"; return context; },
        init: addInitValue,
        build: buildStatement
    };
    return context;
    function addInitValue(value) {
        initExpression = getValueExpression(value);
        return context;
    }
    function buildStatement() {
        return t.variableDeclaration(kind, [
            t.variableDeclarator(t.identifier(name), initExpression)
        ]);
    }
}
exports.declareVar = declareVar;
function declareObjectDestruction(...props) {
    let kind = "const";
    let initExpression;
    const properties = [...props];
    const context = {
        get const() { kind = "const"; return context; },
        get let() { kind = "let"; return context; },
        get var() { kind = "var"; return context; },
        property: addProperty,
        init: addInitValue,
        build: buildStatement
    };
    return context;
    function addProperty(name, as) {
        properties.push({ name, as });
        return context;
    }
    function addInitValue(value) {
        initExpression = getValueExpression(value);
        return context;
    }
    function buildStatement() {
        return t.variableDeclaration(kind, [
            t.variableDeclarator(t.assignmentPattern(t.objectPattern(properties.map(er => t.objectProperty(t.identifier(er.name), t.identifier(er.as ? er.as : er.name), undefined, !er.as))), initExpression))
        ]);
    }
}
exports.declareObjectDestruction = declareObjectDestruction;
function makeCall(name, ...params) {
    const parameters = params.map(getValueExpression);
    const context = {
        param: addParam,
        build: buildStatement
    };
    return context;
    function addParam(...params) {
        parameters.push(...params.map(getValueExpression));
    }
    function buildStatement() {
        return t.callExpression(t.identifier(name), parameters);
    }
}
exports.makeCall = makeCall;
function makeAssignment(target, value) {
    return t.assignmentExpression("=", t.identifier(target), getValueExpression(value));
}
exports.makeAssignment = makeAssignment;
function makeTemplateLiteral(literal) {
    if (typeof literal !== "string") {
        literal = DomUtils_1.getOuterHTML(literal, { decodeEntities: true, xmlMode: true });
    }
    return t.identifier("`" + literal + "`");
}
exports.makeTemplateLiteral = makeTemplateLiteral;
function makeThrow(message) {
    return t.throwStatement(t.newExpression(t.identifier("Error"), [t.stringLiteral(message)]));
}
exports.makeThrow = makeThrow;
function getValueExpression(value) {
    if (value === undefined) {
        return t.identifier("undefined");
    }
    if (isBuilder(value)) {
        value = value.build();
    }
    else if (t.isExpression(value)) {
        return value;
    }
    return t.valueToNode(value);
}
//# sourceMappingURL=builder.js.map