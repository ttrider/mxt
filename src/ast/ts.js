"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ts = __importStar(require("typescript"));
function parseInlineExpressions(content) {
    const state = {
        content,
        hasTokens: false,
        externalReferences: []
    };
    const code = "`" + content + "`";
    const result = getDiagnosticsForText(code);
    if (result) {
        state.hasTokens = true;
        state.externalReferences =
            result
                .filter(r => r.code === 2304)
                .map(r => code.substr(r.start === undefined ? 0 : r.start, r.length));
    }
    return state;
}
exports.parseInlineExpressions = parseInlineExpressions;
;
function getDiagnosticsForText(text) {
    const dummyFilePath = "/file.ts";
    const textAst = ts.createSourceFile(dummyFilePath, text, ts.ScriptTarget.Latest);
    const ht = hasTokens(textAst);
    if (ht) {
        const options = {};
        const host = {
            fileExists: filePath => filePath === dummyFilePath,
            directoryExists: dirPath => dirPath === "/",
            getCurrentDirectory: () => "/",
            getDirectories: () => [],
            getCanonicalFileName: fileName => fileName,
            getNewLine: () => "\n",
            getDefaultLibFileName: () => "",
            getSourceFile: filePath => filePath === dummyFilePath ? textAst : undefined,
            readFile: filePath => filePath === dummyFilePath ? text : undefined,
            useCaseSensitiveFileNames: () => true,
            writeFile: () => { }
        };
        const program = ts.createProgram({
            options,
            rootNames: [dummyFilePath],
            host
        });
        return ts.getPreEmitDiagnostics(program);
    }
}
function hasTokens(node) {
    let value = false;
    traverse(node);
    return value;
    function traverse(node) {
        if (ts.isTemplateSpan(node)) {
            value = true;
            return;
        }
        ts.forEachChild(node, traverse);
    }
}
//# sourceMappingURL=ts.js.map