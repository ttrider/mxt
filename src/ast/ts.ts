import * as ts from "typescript";
import { ExpressionInfo } from "..";

export function parseInlineExpressions(content: string) {

    const state: ExpressionInfo = {

        content,
        externalReferences: []
    };

    const { program, source } = getProgram("`" + content + "`");

    const tokens = getTokens(source);

    if (tokens.length > 0) {
        const diags = ts.getPreEmitDiagnostics(program);

        if (diags) {
            state.tokens = tokens;
            state.externalReferences =
                diags
                    .filter(r => r.code === 2304)
                    .map(r => content.substr(r.start === undefined ? 0 : r.start - 1, r.length));
        }

    }

    return state;
}


function getProgram(content: string) {
    const dummyFilePath = "/file.ts";
    const source = ts.createSourceFile(dummyFilePath, content, ts.ScriptTarget.Latest);

    const options: ts.CompilerOptions = {};
    const host: ts.CompilerHost = {
        fileExists: filePath => filePath === dummyFilePath,
        directoryExists: dirPath => dirPath === "/",
        getCurrentDirectory: () => "/",
        getDirectories: () => [],
        getCanonicalFileName: fileName => fileName,
        getNewLine: () => "\n",
        getDefaultLibFileName: () => "",
        getSourceFile: filePath => filePath === dummyFilePath ? source : undefined,
        readFile: filePath => filePath === dummyFilePath ? content : undefined,
        useCaseSensitiveFileNames: () => true,
        writeFile: () => { }
    };
    const program = ts.createProgram({
        options,
        rootNames: [dummyFilePath],
        host
    });

    return { program, source };
}


function getTokens(node: ts.SourceFile) {

    const tokens: Array<{ start: number, end: number, text: string }> = [];
    traverse(node);
    return tokens;
    function traverse(nd: ts.Node) {
        if (ts.isTemplateLiteral(nd)) {

            const tn = nd as any;

            if (tn.templateSpans) {

                for (const ts of tn.templateSpans) {

                    if (ts.expression) {
                        tokens.push({
                            start: ts.start - 1,
                            end: ts.end - 1,
                            text: ts.text
                        });
                    }
                }
            }
            return;
        }
        ts.forEachChild(nd, traverse);
    }
}

export function generateCode(node: ts.Node | ts.Statement[]) {

    const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });

    if (Array.isArray(node)) {

        const sf = ts.updateSourceFileNode(
            ts.createSourceFile("./dummy.ts", "", ts.ScriptTarget.Latest),
            node);

        return printer.printNode(ts.EmitHint.Unspecified, sf, sf);
    }

    const sf = (!ts.isSourceFile(node)) ? ts.createSourceFile("./dummy.ts", "", ts.ScriptTarget.Latest) : node;
    return printer.printNode(ts.EmitHint.Unspecified, node, sf);
}