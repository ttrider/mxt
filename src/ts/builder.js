"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_1 = __importDefault(require("typescript"));
const DomUtils_1 = require("DomUtils");
function isBuilder(value) {
    return (value && value.build);
}
function isExpression(value) {
    return (value && value._expressionBrand !== undefined);
}
exports.isExpression = isExpression;
function isStatement(value) {
    return (value && value._statementBrand !== undefined);
}
exports.isStatement = isStatement;
function importStatement(from, name) {
    const importMembers = [];
    const context = {
        importMember: (name, as) => {
            importMembers.push({ name, as });
            return context;
        },
        build: buildStatement
    };
    return context;
    function buildStatement() {
        const specifiers = importMembers
            .map(m => {
            if (m.as === undefined || m.as === m.name) {
                return typescript_1.default.createImportSpecifier(undefined, typescript_1.default.createIdentifier(m.name));
            }
            return typescript_1.default.createImportSpecifier(typescript_1.default.createIdentifier(m.name), typescript_1.default.createIdentifier(m.as));
        });
        const cl = typescript_1.default.createImportClause(name === undefined
            ? undefined
            : typescript_1.default.createIdentifier(name), specifiers.length > 0 ? typescript_1.default.createNamedImports(specifiers) : undefined);
        const id = typescript_1.default.createImportDeclaration([], [], cl, typescript_1.default.createStringLiteral(from));
        return id;
    }
}
exports.importStatement = importStatement;
function statements(...node) {
    const items = [];
    const context = {
        add,
        build: buildStatement
    };
    return add(...node);
    function add(...node) {
        for (const statement of node) {
            const st = Array.isArray(statement) ? statement : [statement];
            for (let item of st) {
                if (isBuilder(item)) {
                    item = item.build();
                }
                if (isExpression(item)) {
                    item = typescript_1.default.createStatement(item);
                }
                items.push(item);
            }
        }
        return context;
    }
    function buildStatement() {
        return items;
    }
}
exports.statements = statements;
function declareFunction(functionName) {
    const parameters = [];
    const body = statements();
    let isExport = false;
    const context = {
        get export() { isExport = true; return context; },
        param: addParam,
        body: (...node) => {
            body.add(...node);
            return context;
        },
        build: buildStatement
    };
    return context;
    function buildStatement() {
        // identifiy optional parameters
        let canBeOptional = true;
        for (let i = parameters.length - 1; i >= 0; i--) {
            if (!canBeOptional) {
                parameters[i].questionToken = undefined;
            }
            else {
                if (!parameters[i].questionToken) {
                    canBeOptional = false;
                }
            }
        }
        if (functionName) {
            const modifiers = isExport ? [typescript_1.default.createToken(typescript_1.default.SyntaxKind.ExportKeyword)] : [];
            return typescript_1.default.createFunctionDeclaration([], modifiers, undefined, functionName, undefined, parameters, undefined, typescript_1.default.createBlock(body.build(), true));
        }
        else {
            return typescript_1.default.createArrowFunction(undefined, undefined, parameters, undefined, undefined, typescript_1.default.createBlock(body.build(), true));
        }
    }
    function addParam(name, ...types) {
        const parameter = typescript_1.default.createParameter([], [], undefined, name);
        if (types.length === 0) {
            parameter.type = typescript_1.default.createTypeReferenceNode("any", undefined);
        }
        else if (types.length === 1) {
            parameter.type = getType(types[0]);
        }
        else {
            parameter.type = typescript_1.default.createUnionTypeNode(types.map(getType));
        }
        parameters.push(parameter);
        return context;
        function getType(tp) {
            if (tp === undefined) {
                parameter.questionToken = typescript_1.default.createToken(typescript_1.default.SyntaxKind.QuestionToken);
                return typescript_1.default.createTypeReferenceNode("undefined", undefined);
            }
            if (tp === null) {
                return (typescript_1.default.createNull());
            }
            return typescript_1.default.createTypeReferenceNode(tp, undefined);
        }
    }
}
exports.declareFunction = declareFunction;
function declareVar(name) {
    let kind = typescript_1.default.NodeFlags.Const;
    let initExpression;
    const context = {
        get const() { kind = typescript_1.default.NodeFlags.Const; return context; },
        get let() { kind = typescript_1.default.NodeFlags.Let; return context; },
        get var() { kind = undefined; return context; },
        init: (value) => {
            initExpression = getValueExpression(value);
            return context;
        },
        build: buildStatement
    };
    return context;
    function buildStatement() {
        return typescript_1.default.createVariableStatement([], typescript_1.default.createVariableDeclarationList([
            typescript_1.default.createVariableDeclaration(name, undefined, initExpression),
        ], kind));
    }
}
exports.declareVar = declareVar;
function declareObjectDestruction(...props) {
    let kind = typescript_1.default.NodeFlags.Const;
    let initExpression;
    const properties = [...props];
    const context = {
        get const() { kind = typescript_1.default.NodeFlags.Const; return context; },
        get let() { kind = typescript_1.default.NodeFlags.Let; return context; },
        get var() { kind = undefined; return context; },
        property: (name, as) => {
            properties.push({ name, as });
            return context;
        },
        init: (value) => {
            initExpression = getValueExpression(value);
            return context;
        },
        build: buildStatement
    };
    return context;
    function buildStatement() {
        return typescript_1.default.createVariableStatement([], typescript_1.default.createVariableDeclarationList([
            typescript_1.default.createVariableDeclaration(typescript_1.default.createObjectBindingPattern(properties.map(item => {
                if (item.as === undefined) {
                    return typescript_1.default.createBindingElement(undefined, undefined, item.name);
                }
                return typescript_1.default.createBindingElement(undefined, item.name, item.as);
            })), undefined, initExpression),
        ], kind));
    }
}
exports.declareObjectDestruction = declareObjectDestruction;
function makeCall(name, ...params) {
    const parameters = params.map(getValueExpression);
    const context = {
        param: (...params) => {
            parameters.push(...params.map(getValueExpression));
        },
        build: buildStatement
    };
    return context;
    function buildStatement() {
        return typescript_1.default.createCall(typescript_1.default.createIdentifier(name), undefined, parameters);
    }
}
exports.makeCall = makeCall;
function makeAssignment(target, value) {
    return typescript_1.default.createAssignment(typescript_1.default.createIdentifier(target), getValueExpression(value));
}
exports.makeAssignment = makeAssignment;
function makeTemplateLiteral(literal) {
    if (typeof literal !== "string") {
        literal = DomUtils_1.getOuterHTML(literal, { decodeEntities: true, xmlMode: true });
    }
    return typescript_1.default.createIdentifier("`" + literal + "`");
}
exports.makeTemplateLiteral = makeTemplateLiteral;
function makeThrow(message) {
    return typescript_1.default.createThrow(typescript_1.default.createNew(typescript_1.default.createIdentifier("Error"), undefined, [typescript_1.default.createStringLiteral(message)]));
}
exports.makeThrow = makeThrow;
function getValueExpression(value) {
    if (value === undefined) {
        return typescript_1.default.createIdentifier("undefined");
    }
    if (value === null) {
        return typescript_1.default.createNull();
    }
    if (isBuilder(value)) {
        value = value.build();
    }
    if (value.kind !== undefined) {
        return value;
    }
    return typescript_1.default.createLiteral(value);
}
function generateCode(node) {
    const sf = (!typescript_1.default.isSourceFile(node)) ? typescript_1.default.createSourceFile("./dummy.ts", "", typescript_1.default.ScriptTarget.Latest) : node;
    const printer = typescript_1.default.createPrinter({ newLine: typescript_1.default.NewLineKind.LineFeed });
    return printer.printNode(typescript_1.default.EmitHint.Unspecified, node, sf);
}
exports.generateCode = generateCode;
//# sourceMappingURL=builder.js.map