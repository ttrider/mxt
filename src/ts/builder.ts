import ts from "typescript";






export function importStatement(from: string, name?: string) {

    const importMembers: Array<{ name: string, as?: string }> = [];

    const context = {

        importMember: (name: string, as?: string) => {
            importMembers.push({ name, as });
            return context;
        },

        build: buildStatement
    };

    return context;

    function buildStatement() {

        const specifiers = importMembers
            .map(m => {

                if (m.as === undefined) {
                    return ts.createImportSpecifier(undefined, ts.createIdentifier(m.name));
                }
                return ts.createImportSpecifier(ts.createIdentifier(m.name), ts.createIdentifier(m.as));
            });

        const cl = ts.createImportClause(
            name === undefined
                ? undefined
                : ts.createIdentifier(name),
            specifiers.length > 0 ? ts.createNamedImports(specifiers) : undefined);

        const id = ts.createImportDeclaration([], [], cl, ts.createStringLiteral(from));

        return id;
    }
}


interface IBuilder {
    build(): ts.Statement
}

function isBuilder(value: any): value is IBuilder {
    return (value && value.build);
}

export declare type StatementItem =
    ts.Statement |
    IBuilder
    ;


export function statements(...node: Array<StatementItem | StatementItem[]>) {

    const items: ts.Statement[] = [];

    const context =
    {
        add,

        build: buildStatement
    };

    return add(...node);

    function add(...node: Array<StatementItem | StatementItem[]>) {
        for (const statement of node) {

            const st = Array.isArray(statement) ? statement : [statement];

            for (let item of st) {
                if (isBuilder(item)) {
                    item = item.build();
                }
                items.push(item);
            }
        }
        return context;
    }

    function buildStatement() {
        return items;
    }
}

export function declareFunction(functionName?: string) {

    const parameters: ts.ParameterDeclaration[] = [];
    const body = statements();
    let isExport = false;

    const context = {
        get export() { isExport = true; return context },


        param: addParam,
        body: (...node: Array<StatementItem | StatementItem[]>) => {
            body.add(...node);
            return context;
        },

        build: buildStatement
    };

    return context;

    function buildStatement() {

        // identifiy optional parameters
        let canBeOptional = true;
        for (let i = parameters.length - 1; i >= 0; i--) {
            if (!canBeOptional) {
                parameters[i].questionToken = undefined;
            } else {
                if (!parameters[i].questionToken) {
                    canBeOptional = false;
                }
            }
        }


        if (functionName) {
            const modifiers = isExport ? [ts.createToken(ts.SyntaxKind.ExportKeyword)] : [];
            return ts.createFunctionDeclaration([], modifiers, undefined, functionName, undefined, parameters, undefined, ts.createBlock([]));
        } else {
            return ts.createArrowFunction(undefined, undefined, parameters, undefined, undefined, ts.createBlock([]));
        }
    }

    function addParam(name: string, ...types: (null | string | undefined)[]) {

        const parameter = ts.createParameter(
            [],
            [],
            undefined,
            name,
        );

        if (types.length === 0) {
            parameter.type = ts.createTypeReferenceNode("any", undefined);
        } else if (types.length === 1) {
            parameter.type = getType(types[0]);
        } else {
            parameter.type = ts.createUnionTypeNode(types.map(getType));
        }

        parameters.push(parameter);

        return context;

        function getType(tp: null | string | undefined) {

            if (tp === undefined) {
                parameter.questionToken = ts.createToken(ts.SyntaxKind.QuestionToken);
                return ts.createTypeReferenceNode("undefined", undefined)
            }

            if (tp === null) {
                return (ts.createNull());
            }

            return ts.createTypeReferenceNode(tp, undefined);
        }
    }
}

export function declareVar(name: string) {

    let kind: ts.NodeFlags.Const | ts.NodeFlags.Let | undefined = ts.NodeFlags.Const;
    let initExpression: ts.Expression;

    const context = {
        get const() { kind = ts.NodeFlags.Const; return context },
        get let() { kind = ts.NodeFlags.Let; return context },
        get var() { kind = undefined; return context },

        init: (value: any) => {
            initExpression = getValueExpression(value);
            return context;
        },

        build: buildStatement
    };

    return context;

    function buildStatement() {

        return ts.createVariableStatement(
            [],
            ts.createVariableDeclarationList(
                [
                    ts.createVariableDeclaration(
                        name,
                        undefined,
                        initExpression,
                    ),
                ],
                kind
            ),
        )

    }
}




function getValueExpression(value?: any) {
    if (value === undefined) {
        return ts.createIdentifier("undefined");
    }

    if (value === null) {
        return ts.createNull();
    }

    if (isBuilder(value)) {
        value = value.build();
    }

    if (value.kind !== undefined) {
        return value;
    }

    return ts.createLiteral(value)

}

export function generateCode(node: ts.Node) {

    const sf = (!ts.isSourceFile(node)) ? ts.createSourceFile("./dummy.ts", "", ts.ScriptTarget.Latest) : node;

    const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
    return printer.printNode(ts.EmitHint.Unspecified, node, sf);

}