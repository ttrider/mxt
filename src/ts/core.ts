import ts from "typescript";
export function isExpression(node?: ts.Node): node is ts.Expression {

    return node !== undefined && (
        ts.isArrayLiteralExpression(node)
        || ts.isObjectLiteralExpression(node)
        || ts.isPropertyAccessExpression(node)
        || ts.isElementAccessExpression(node)
        || ts.isCallExpression(node)
        || ts.isNewExpression(node)
        || ts.isTaggedTemplateExpression(node)
        || ts.isParenthesizedExpression(node)
        || ts.isFunctionExpression(node)
        || ts.isDeleteExpression(node)
        || ts.isTypeOfExpression(node)
        || ts.isVoidExpression(node)
        || ts.isAwaitExpression(node)
        || ts.isPrefixUnaryExpression(node)
        || ts.isPostfixUnaryExpression(node)
        || ts.isBinaryExpression(node)
        || ts.isConditionalExpression(node)
        || ts.isTemplateExpression(node)
        || ts.isYieldExpression(node)
        || ts.isClassExpression(node)
        || ts.isOmittedExpression(node)
        || ts.isAsExpression(node)
        || ts.isNonNullExpression(node)
        || ts.isJsxExpression(node)
        || ts.isJSDocTypeExpression(node)
        || ts.isLiteralExpression(node)
        || ts.isCallLikeExpression(node)
        || ts.isCallOrNewExpression(node)
        || ts.isAssertionExpression(node)
        || ts.isArrowFunction(node)
    );
}
export function isStatement(node: ts.Node): node is ts.Statement {

    return node !== undefined && (
        ts.isVariableStatement(node)
        || ts.isEmptyStatement(node)
        || ts.isExpressionStatement(node)
        || ts.isIfStatement(node)
        || ts.isDoStatement(node)
        || ts.isWhileStatement(node)
        || ts.isForStatement(node)
        || ts.isForInStatement(node)
        || ts.isForOfStatement(node)
        || ts.isContinueStatement(node)
        || ts.isBreakStatement(node)
        || ts.isBreakOrContinueStatement(node)
        || ts.isReturnStatement(node)
        || ts.isWithStatement(node)
        || ts.isSwitchStatement(node)
        || ts.isLabeledStatement(node)
        || ts.isThrowStatement(node)
        || ts.isTryStatement(node)
        || ts.isDebuggerStatement(node)
    );

}


export declare type NodeArrayBuilder<T extends ts.Node> = ts.NodeArray<T> & {
    add: (...values: T[]) => NodeArrayBuilder<T>
};

export function fromNodeArray<T extends ts.Node>(nodeArray: ts.NodeArray<T> | undefined, ...values: T[]) {

    const list = (nodeArray === undefined ? [] : nodeArray) as NodeArrayBuilder<T>;

    list.add = (...values: T[]) => {
        (list as any).push(...values);
        return list;
    }

    return list;
}


declare type ModifiersKeywords = ts.SyntaxKind.AbstractKeyword |
    ts.SyntaxKind.AsyncKeyword |
    ts.SyntaxKind.ConstKeyword |
    ts.SyntaxKind.DeclareKeyword |
    ts.SyntaxKind.DefaultKeyword |
    ts.SyntaxKind.ExportKeyword |
    ts.SyntaxKind.PublicKeyword |
    ts.SyntaxKind.PrivateKeyword |
    ts.SyntaxKind.ProtectedKeyword |
    ts.SyntaxKind.ReadonlyKeyword |
    ts.SyntaxKind.StaticKeyword;

export declare type ModifiersBuilder = ts.ModifiersArray & {
    add: (...values: ModifiersKeywords[]) => ModifiersBuilder
}
export function modifiersBuilder(host: { modifiers?: ts.ModifiersArray | undefined }) {

    const modifiers = (host.modifiers === undefined ? [] : host.modifiers) as ModifiersBuilder;

    modifiers.add = (...values: ModifiersKeywords[]) => {

        for (const kind of values) {
            if (!modifiers.find(t => t.kind === kind)) {
                (modifiers as any).push(ts.createModifier(kind));
            }
        }
        return modifiers;
    }

    return modifiers;

}
//modifiers: ts.ModifiersArray | undefined