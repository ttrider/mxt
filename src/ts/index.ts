import ts from "typescript";

export * from "./block";
export * from "./core";
export * from "./declare";
export * from "./for";
export * from "./function";
export * from "./import";
export * from "./literal";
export * from "./parameter";
export * from "./type";


export function generateCode(node: ts.Node) {

    const sf = (!ts.isSourceFile(node)) ? ts.createSourceFile("./dummy.ts", "", ts.ScriptTarget.Latest) : node;

    const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
    return printer.printNode(ts.EmitHint.Unspecified, node, sf);

}


