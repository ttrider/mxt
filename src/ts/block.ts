import ts, { isBlock } from "typescript";
import { isExpression } from "./core";

export declare type BlockArgs = Array<ts.Statement | ts.Expression | ts.Block | undefined>;


declare type StatementListBuilder = ts.Statement[] & {
    add: (...statement: BlockArgs) => StatementListBuilder;
    toBlock: (multiLine?: boolean | undefined) => ts.Block;

}
export function StatementList(...statement: BlockArgs) {

    const statements: StatementListBuilder = convert(statement) as StatementListBuilder;

    statements.toBlock = (multiLine?: boolean | undefined) => {
        return ts.createBlock(statements, multiLine);
    }

    statements.add = (...statement: BlockArgs) => {
        statements.push(...convert(statement));
        return statements;
    }

    return statements;

    function convert(statements: BlockArgs) {

        return statements.reduce<ts.Statement[]>(
            (ret, item) => {

                if (item !== undefined) {
                    if (isBlock(item)) {
                        ret.push(...item.statements);
                    } else {
                        ret.push(isExpression(item) ? ts.createStatement(item) : item);
                    }
                }
                return ret;
            }, []
        );
    }
}


declare type BlockBuilder = ts.Block & {
    add: (...statement: BlockArgs) => BlockBuilder
};

export function Block(...statement: BlockArgs) {

    let obj = ts.createBlock([], true) as BlockBuilder;
    (obj as any).statements = StatementList(...statement);
    obj.add = (obj as any).statements.add;

    return obj;

}