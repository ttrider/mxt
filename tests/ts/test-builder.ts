import { importStatement, generateCode } from "../../src/ts/builder";


describe("builder", () => {
    test("import { member } from fromModule", () => {

        const is = importStatement("fromModule")
            .importMember("member")
            .build();

        const istext = generateCode(is);
        expect(istext).toMatchSnapshot();
    });

    test("import { member as local } from fromModule", () => {

        const is = importStatement("fromModule")
            .importMember("member", "local")
            .build();

        const istext = generateCode(is);
        expect(istext).toMatchSnapshot();
    });

    test("import defname from fromModule", () => {

        const is = importStatement("fromModule", "defname")
            .build();

        const istext = generateCode(is);
        expect(istext).toMatchSnapshot();
    });

    test("import defmember, { member1 as local1, member 2 } from fromModule", () => {

        const is = importStatement("fromModule", "defmember")
            .importMember("member1", "local1")
            .importMember("member2")
            .build();

        const istext = generateCode(is);

        expect(istext).toMatchSnapshot();
    });

});