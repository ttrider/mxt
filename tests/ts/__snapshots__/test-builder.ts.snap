// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder import { member } from fromModule 1`] = `"import { member } from \\"fromModule\\";"`;

exports[`builder import { member as local } from fromModule 1`] = `"import { member as local } from \\"fromModule\\";"`;

exports[`builder import defmember, { member1 as local1, member 2 } from fromModule 1`] = `"import defmember, { member1 as local1, member2 } from \\"fromModule\\";"`;

exports[`builder import defname from fromModule 1`] = `"import defname from \\"fromModule\\";"`;
